services/modules/discord/tasks.py

```diff
@@ -0,0 +1,132 @@
+        :return: bool
+        """
+        try:
+            user.discord
+        except ObjectDoesNotExist:
+            return False
+        else:
+            return True
+
+    @staticmethod
+    @app.task(bind=True)
+    @only_one(key="Discord", timeout=60 * 5)
+    def update_groups(task_self, pk):
+        user = User.objects.get(pk=pk)
+        logger.debug("Updating discord groups for user %s" % user)
+        if DiscordTasks.has_account(user):
+            groups = []
+            for group in user.groups.all():
+                groups.append(str(group.name))
+            if len(groups) == 0:
+                logger.debug("No syncgroups found for user. Adding empty group.")
+                groups.append('empty')
+            logger.debug("Updating user %s discord groups to %s" % (user, groups))
+            try:
+                DiscordOAuthManager.update_groups(user.discord.uid, groups)
+            except Exception as e:
+                if task_self:
+                    logger.exception("Discord group sync failed for %s, retrying in 10 mins" % user)
+                    raise task_self.retry(countdown=60 * 10)
+                else:
+                    # Rethrow
+                    raise e
+            logger.debug("Updated user %s discord groups." % user)
+        else:
+            logger.debug("User does not have a discord account, skipping")
+
+    @staticmethod
+    @app.task
+    def update_all_groups():
+        logger.debug("Updating ALL discord groups")
+        for discord_user in DiscordUser.objects.exclude(uid__exact=''):
+            DiscordTasks.update_groups.delay(discord_user.user.pk)
+
+    @staticmethod
+    @app.task(bind=True)
+    def update_nickname(self, pk):
+        user = User.objects.get(pk=pk)
+        logger.debug("Updating discord nickname for user %s" % user)
+        if DiscordTasks.has_account(user):
+            character = EveManager.get_main_character(user)
+            logger.debug("Updating user %s discord nickname to %s" % (user, character.character_name))
+            try:
+                DiscordOAuthManager.update_nickname(user.discord.uid, character.character_name)
+            except Exception as e:
+                if self:
+                    logger.exception("Discord nickname sync failed for %s, retrying in 10 mins" % user)
+                    raise self.retry(countdown=60 * 10)
+                else:
+                    # Rethrow
+                    raise e
+            logger.debug("Updated user %s discord nickname." % user)
+        else:
+            logger.debug("User %s does not have a discord account" % user)
+
+    @staticmethod
+    @app.task
+    def update_all_nicknames():
+        logger.debug("Updating ALL discord nicknames")
+        for discord_user in DiscordUser.objects.exclude(uid__exact=''):
+            DiscordTasks.update_nickname.delay(discord_user.user.user_id)
+
+    @classmethod
+    def disable(cls):
+        if settings.ENABLE_AUTH_DISCORD:
```

Comment:
I would go further than this and say that the `ENABLE` settings should maybe not be settings at all and rather base services access/permissions on groups. Say, for instance, have a `OneToOneField` on groups per service, and then on a user you can do `user.groups.filter(example_enabled__isnull=False)` (or something to that effect). This has the added benefit of increased service access granularity for users who want or need it.

Alternatively maybe the setting should be a dictionary of `states: bool`. Otherwise every time you add a new state you gain an extra 11-13 settings (11 services now, 13 soon) and that may get out of hand fast, depending on how many states are intended to be added.

Either way I think this change should be taken to a future PR.